src/sentry/tasks/deletion/hybrid_cloud.py

```diff
@@ -0,0 +1,214 @@
+import sentry_sdk
+from django.apps import apps
+from django.db.models import Manager, Max
+
+from sentry.db.models.fields.hybrid_cloud_foreign_key import HybridCloudForeignKey
+from sentry.models import TombstoneBase
+from sentry.silo import SiloMode
+from sentry.tasks.base import instrumented_task
+from sentry.utils import json, metrics, redis
+
+
+def deletion_silo_modes() -> List[SiloMode]:
+    cur = SiloMode.get_current_mode()
+    result: List[SiloMode] = []
+    if cur != SiloMode.REGION:
+        result.append(SiloMode.CONTROL)
+    if cur != SiloMode.CONTROL:
+        result.append(SiloMode.REGION)
+    return result
+
+
+def get_watermark_key(prefix: str, field: HybridCloudForeignKey) -> str:
+    return f"{prefix}.{field.model._meta.db_table}.{field.name}"
+
+
+def get_watermark(prefix: str, field: HybridCloudForeignKey) -> Tuple[int, str]:
+    with redis.clusters.get("default").get_local_client_for_key("deletions.watermark") as client:
+        key = get_watermark_key(prefix, field)
+        v = client.get(key)
+        if v is None:
+            result = (0, uuid4().hex)
+            client.set(key, json.dumps(result))
+            return result
+        return tuple(json.loads(v))
+
+
+def set_watermark(
+    prefix: str, field: HybridCloudForeignKey, value: int, prev_transaction_id: str
+) -> None:
+    with redis.clusters.get("default").get_local_client_for_key("deletions.watermark") as client:
+        client.set(
+            get_watermark_key(prefix, field),
+            json.dumps((value, sha1(prev_transaction_id.encode("utf8")).hexdigest())),
+        )
+    metrics.gauge(
+        "deletion.hybrid_cloud.low_bound",
+        value,
+        tags=dict(
+            table_name=field.model._meta.db_table,
+            field_name=field.name,
+            watermark=prefix,
+        ),
+    )
+
+
+def chunk_watermark_batch(
+    prefix: str, field: HybridCloudForeignKey, manager: Manager, *, batch_size: int
+) -> Tuple[int, int, bool, str]:
+    lower, transaction_id = get_watermark(prefix, field)
+    upper = manager.aggregate(Max("id"))["id__max"] or 0
+    batch_upper = min(upper, lower + batch_size)
+    return lower, upper, batch_upper + batch_size < upper, transaction_id
+
+
+@instrumented_task(
+    name="sentry.tasks.deletion.hybrid_cloud.schedule_hybrid_cloud_foreign_key_jobs",
+    queue="cleanup",
+    acks_late=True,
+)
+def schedule_hybrid_cloud_foreign_key_jobs():
+    for silo_mode in deletion_silo_modes():
+        for app, app_models in apps.all_models.items():
+            for model in app_models.values():
+                if not hasattr(model._meta, "silo_limit"):
+                    continue
+
+                # Only process models local this operational mode.
+                if silo_mode not in model._meta.silo_limit.modes:
+                    continue
+
+                for field in model._meta.fields:
+                    if not isinstance(field, HybridCloudForeignKey):
+                        continue
```

Comment:
Should `HybridCloudForeignKey` keep state of where it has been used so that this quadruple loop can be smaller?